asset-transfer-ledger-queries/application-javascript/app.js

```diff
@@ -0,0 +1,213 @@
+ *   -- How to query and check the results
+ *
+ * To see the SDK workings, try setting the logging to show on the console before running
+ *        export HFC_LOGGING='{"debug":"console"}'
+ */
+async function main() {
+	try {
+		// build an in memory object with the network configuration (also known as a connection profile)
+		const ccp = buildCCP();
+
+		// build an instance of the fabric ca services client based on
+		// the information in the network configuration
+		const caClient = buildCAClient(FabricCAServices, ccp);
+
+		// setup the wallet to hold the credentials of the application user
+		const wallet = await buildWallet(Wallets, walletPath);
+
+		// in a real application this would be done on an administrative flow, and only once
+		await enrollAdmin(caClient, wallet);
+
+		// in a real application this would be done only when a new user was required to be added
+		// and would be part of an administrative flow
+		await registerUser(caClient, wallet, userId, 'org1.department1');
+
+		// Create a new gateway instance for interacting with the fabric network.
+		// In a real application this would be done as the backend server session is setup for
+		// a user that has been verified.
+		const gateway = new Gateway();
+
+		try {
+			// setup the gateway instance
+			// The user will now be able to create connections to the fabric network and be able to
+			// submit transactions and query. All transactions submitted by this gateway will be
+			// signed by this user using the credentials stored in the wallet.
+			await gateway.connect(ccp, {
+				wallet,
+				identity: userId,
+				discovery: {enabled: true, asLocalhost: true} // using asLocalhost as this gateway is using a fabric network deployed locally
+			});
+
+			// Build a network instance based on the channel where the smart contract is deployed
+			const network = await gateway.getNetwork(channelName);
+
+			// Get the contract from the network.
+			const contract = network.getContract(chaincodeName);
+
+			// Initialize a set of asset data on the channel using the chaincode 'InitLedger' function.
+			// This type of transaction would only be run once by an application the first time it was started after it
+			// deployed the first time. Any updates to the chaincode deployed later would likely not need to run
+			// an "init" type function.
+			try {
+				console.log('\n--> Submit Transaction: InitLedger, function creates the initial set of assets on the ledger');
+				await contract.submitTransaction('InitLedger');
+				console.log('*** Result: committed');
+			} catch(initError) {
+				// this is error is OK if we are rerunning this app without restarting
+				console.log(`initLedger failed :: ${initError}`)
+			}
+
+
+			// Let's try a query type operation (function).
+			// This will be sent to just one peer and the results will be shown.
+			console.log('\n--> Evaluate Transaction: GetAssetsByRange, function returns all the current assets on the ledger');
```

Comment:
returns a range of asset keys on the ledger.
Note, you can use empty string for startKey and endKey to do an open-ended query of all assets.

Reply:
add a sample call showing all four options for using the query
- start open end open
- start open end specific
- start specific end open
- start specific end specific